Fix the instructions for how to run a specific system test suite (#11692)
microsoft/vscode-python: CONTRIBUTING.md

1. Run the tests via `npm run` or the Debugger launch options (you can "Start Without Debugging").
1. **Note** you will be running tests under the default Python interpreter for the system.

You can also run the tests from the command-line (after compiling):

```shell
npm run testSingleWorkspace  # will launch the VSC UI
npm run testMultiWorkspace  # will launch the VSC UI
```

> #### Customising the test run

Reviewer: ```suggestion
#### Customising the Test Run
```

If you want to change which tests are run or which version of Python is used,

> you can do this by setting environment variables.

Reviewer: Let's be sure to mention where to set the environment variables:

```suggestion
you can do this by setting environment variables, whether running tests
from the command line or in VS Code.
```

Author: 6ca52c675

* Setting `CI_PYTHON_PATH` lets you change the version of python the tests are executed with
* Setting `VSC_PYTHON_CI_TEST_GREP` lets you filter the tests by name

_`CI_PYTHON_PATH`_

By default tests will be executed using the default interpreter for the
environment they are launched from. When launching from VSCode this will be
whatever that the default `python` i for your local machine. When launching from

> the command line this will be the version of `python` the shell is using.

Author: Ah, yeah, that's much clearer :+1:

> the command line this will be the version of `python` the shell is using.

Author: 649bce3df

This environment variable allows setting a specific python interpreter by

> specifying the absolute path to the `python` binary.

Author: Ah, I hadn't expected that this variable could be used to specify the search name -- only that it needed path. That's a useful distinction to have.

This is also the mechanism for testing against other versions of Python.

_`VSC_PYTHON_CI_TEST_GREP`_

This environment variable allows providing a pattern which will be matched
against test names to be run. By default all tests are run.

> For example, to run only the tests in the `Sorting` suite (from
> `src/test/format/extension.sort.test.ts`) you would set the value to `Sorting`.

Reviewer: This is a good example.  It would be worth adding a link:

```suggestion
For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`](https://github.com/Microsoft/vscode-python/blob/b328ba12331ed34a267e32e77e3e4b1eff235c13/src/test/format/extension.sort.test.ts))
you would set the value to `Sorting`.
```

and it may be worth adding a couple more examples to be sure things are clear:

```suggestion
For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`](https://github.com/Microsoft/vscode-python/blob/b328ba12331ed34a267e32e77e3e4b1eff235c13/src/test/format/extension.sort.test.ts))
you would set the value to `Sorting`.  If you used a value of `FileSystem`
then 15 different suites from 9 different test files would match, though
only a subset of those would run depending on the kind of tests you
are running (e.g. unit, functional).  A value of `.*error` would get you
87 tests and 2 suites across many different files.
```

Author: I can see what you're getting at with the examples, however I'm concerned that the numbers would get out of date fairly quickly. I'm also not sure that those example numbers are quite right. As this subsection relates only to the running of the system tests, I wouldn't expect any of the unit or functional tests to run anyway, meaning that only the three suites from `src/test/common/platform/filesystem.functional.test.ts` would get run. Does that match your expectation of the behaviour?

Reviewer: Good point.  The numbers aren't critical.  The worth of the example is in demonstrating the scope of the different `--grep` values.  Let's not stress too much about the numbers, but feel free to approach the example differently if you think you can do so without impacting the main point.

Author: 14966dc27 adds a regex example which covers tests from multiple suites.

Be sure to escape any grep-sensitive characters in your suite name.

_Launching from VSCode_

> In order to set these values when launching the tests from VSCode you should

Reviewer: small edit ("values" -> "env vars"):

```suggestion
In order to set these env vars when launching the tests from VSCode you should
```

Author: f64644629

edit the `launch.json` file. For example you can add the following to the
appropriate configuration you want to run to change the interpreter used during
testing:

```js
    "env": {
        "CI_PYTHON_PATH": "/absolute/path/to/interpreter/of/choice/python"
    }
```

> 

Reviewer: I'm pretty sure there are still corner cases where "system" tests don't respect the environment variables, so let's keep at least a mention of that:

```suggestion

In some rare cases in the "system" tests the `VSC_PYTHON_CI_TEST_GREP`
environment variable is ignored.  If that happens then you will need to temporarily
modify the "const grep = " line in [`src/test/index.ts`](https://github.com/Microsoft/vscode-python/blob/b328ba12331ed34a267e32e77e3e4b1eff235c13/src/test/index.ts#L21).

```

Author: ee0f5e052

this will almost always mean Python 2 is used, which probably isn't what you
want.

By setting the `CI_PYTHON_PATH` environment variable you can
control the exact Python executable that gets used.  If the executable
you specify isn't on `$PATH` then be sure to use an absolute path.

This is also the mechanism for testing against other versions of Python.

_`VSC_PYTHON_CI_TEST_GREP`_

This environment variable allows providing a regular expression which will
be matched against suite and test "names" to be run. By default all tests
are run.

For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`][extension.sort.test.ts]) you would
set the value to `Sorting`.

Be sure to escape any grep-sensitive characters in your suite name.

In some rare cases in the "system" tests the `VSC_PYTHON_CI_TEST_GREP`
environment variable is ignored.  If that happens then you will need to
temporarily modify the `const grep = ` line in
[`src/test/index.ts`][src/test/index.ts].

[extension.sort.test.ts]: https://github.com/Microsoft/vscode-python/blob/84f9c7a174111e8826a6f06832a7c906c60eadad/src/test/format/extension.sort.test.ts

> [src/test/index.ts]: https://github.com/Microsoft/vscode-python/blob/84f9c7a174111e8826a6f06832a7c906c60eadad/src/test/index.ts#L64

Reviewer: Any reason not to just in-line these links?

Author: Only that they're rather long and make it harder to read the markdown.

Reviewer: I say don't worry about it.  I brought it up because the disconnected links felt a bit disjointed and I don't think in-lining will be a problem for readers.  FWIW, in cases like this I usually wrap the lines so the long link is on its own line:

```markdown
For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`](https://github.com/Microsoft/vscode-python/blob/84f9c7a174111e8826a6f06832a7c906c60eadad/src/test/format/extension.sort.test.ts)
) you would set the value to `Sorting`.
```

It might also be a bit easier for readers if you use a shorter ref hash:

```markdown
For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`](https://github.com/Microsoft/vscode-python/blob/84f9c7a174/src/test/format/extension.sort.test.ts)
) you would set the value to `Sorting`.
```

or just "master":

```markdown
For example, to run only the tests in the `Sorting` suite (from
[`src/test/format/extension.sort.test.ts`](https://github.com/Microsoft/vscode-python/blob/master/src/test/format/extension.sort.test.ts)
) you would set the value to `Sorting`.
```

Author: Good thought on using a shorter hash. I'd rather not use `master` as that's a floating reference meaning that changes to the target file are likely to make the line number point to the wrong place.

Regarding inlining: I like the idea of wrapping the lines as you suggest, though in this case it would result in an extra space appearing between the link and the closing paren, which would then read weirdly in the rendered form.

Reviewer: At this point we're close enough and the question of inlining isn't significant enough to hold things up. :slightly_smiling_face:

Author: oh, sorry, I should have posted -- I did inline them in the end (https://github.com/microsoft/vscode-python/pull/11692/commits/faaaec8b79930f99d01a82a33fb01f806f2aec23)
